### ISSUE TITLE
Dependant keys

### ISSUE THREAD

**Issue by antulik** [orgsync/active_interaction#376]
This is an idea/request for the feature. Would be nice to have ability to define dependant keys. Let's assume example

```ruby
class Update < ActiveInteraction::Base
  integer :user_id
  object :user, default: -> { User.find(user_id) }

  def execute
  end
end
```

I've found it's very convenient to define both `user_id` and `user`. In code you would use `user_id` as it is required for using class as a background job. 

In tests you can pass `user` object directly and don't need to mock anything. Which makes it extremely easy to test.

So you should pass either `user_id` or `user`. But not both and exactly one of them. Right now `active_interaction` does not allow that. 

As a proposed solution

```ruby
class Update < ActiveInteraction::Base
  integer :user_id
  object :user, default: -> { User.find(user_id) }, depends_on: :user_id

  # or

  integer :user_id, provides: :user
  object :user, default: -> { User.find(user_id) }


  def execute
  end
end
```

in the example `user_id` will be marked as virtual and only `user` will be available in the code.

This is just an idea and I'd like to see what others think as it worked very well for me so far.

**Comment by tfausak** [comment]
This is an interesting idea! I have a couple thoughts about it.

I feel your pain about using IDs with background jobs. At OrgSync, one of the way we get around that is to have a worker whose sole job is to find all the records and hand them off to an interaction. So if you had an interaction that required a `User`, you would pass that user's ID to your worker. Then your worker would so something like this:

``` rb
UserInteraction.run!(user: User.find(user_id))
```

If you want to do something like this with ActiveInteraction today, you can get most of the way there. Ever since #332 we set up the default's binding to be the interaction itself. So you can write a `user` input that depends on `user_id` and vice versa.

``` rb
User = Struct.new(:id)

class Example < ActiveInteraction::Base
  object :user, default: -> { User.new(user_id) if user_id? }
  integer :user_id, default: -> { user.id if user? }

  def execute
    inputs
  end
end

Example.run!(user_id: 12)
# => {:user=>#<struct User id=12>, :user_id=>12}
Example.run!(user: User.new(34))
# => {:user=>#<struct User id=34>, :user_id=>34}
```

This has a couple down sides that aren't immediately obvious. If you pass *both* `user` and `user_id`, they may not refer to the same user. 

``` rb
Example.run!(user_id: 12, user: User.new(34))
# => {:user=>#<struct User id=34>, :user_id=>12}
```

Also since both values have a default, you don't actually have to pass either of them.

``` rb
Example.run!
# => {:user=>nil, :user_id=>nil}
```

Both of these problems can be solved with additional validation, but then you're doing a lot of work manually. One way around that is to define your own helper (`object_or_id` perhaps) that sets all of the inputs and validations up for you. 

I'm not opposed to adding something like this to ActiveInteraction, but it's not something that comes up too often for me.

**Comment by antulik** [comment]
@tfausak  I like your idea about `object_or_id`. Have two more thoughts about it:

1. Add support of [GlobalID](https://github.com/rails/globalid), so `user` will accept object or string with global id format.

2. Finder option:

```ruby
class Example < ActiveInteraction::Base
  object :user, finder: -> { User.new(user_id) }

  def execute
  end
end
```
Problem with this type of`user_id` is unknown or should be assumed as `integer`. Also `user_id` will be present in the class, which is also not ideal.

Thoughts?

**Comment by voondo** [comment]
+1 for GlobalID support

**Comment by tfausak** [comment]
I hadn't heard of GlobalID before. Thanks for introducing me to that!

I made a proof of concept for this feature. It convinced me that this shouldn't be in ActiveInteraction. There is simply too much stuff that needs to be configured. 

The way we generally solve problems like this at OrgSync is to create a concern that we mix into the interactions that need this behavior. Here's what this one might look like:

``` rb
require 'active_interaction' # 3.2.0

User = Struct.new(:id)

module ObjectOrId
  extend ActiveSupport::Concern

  class_methods do
    # This assumes that you want an input called `name` and another called
    # `name_id`. Maybe you want a different suffix, in which case you'd need to
    # either add an option or change this code.
    def object_or_id(name)
      # If you were using something like GlobalID, this would have to be a
      # string (or even an object with `class: GlobalID`) instead of an
      # integer.
      integer :"#{name}_id",
        default: lambda {
          if given?(name)
            object = public_send(name)
            # This part would depend on the semantics of your application. It
            # might be `object.uuid` or `object.to_global_id`.
            object.id
          end
        }

      object name,
        default: lambda { |input|
          if given?("#{name}_id")
            # This reads the object's class from the input itself. However this
            # proof of concept doesn't allow changing the class. That would
            # require passing some options to `object_or_id`.
            klass = input.send(:klass)
            id = public_send("#{name}_id")
            # This part would also depend on the semantics of your application.
            # It might be `klass.find(id)` or `GlobalId::Locator.locate(id)`.
            klass.new(id)
          end
        }

      # This validation makes sure that at least the object or ID was given. If
      # you want the input to be optional, you don't need this.
      validate do
        if !given?("#{name}_id") && !given?(name)
          errors.add(:base, "must give one of #{name}_id or #{name}")
        end
      end

      # This validation makes sure that either the object was given or the ID
      # was given, but not both. It may be that you're fine with both being
      # given. Or it's ok as long as they're the same.
      validate do
        if given?("#{name}_id") && given?(name)
          errors.add(:base, "can't give both #{name}_id and #{name}")
        end
      end
    end
  end
end

class Example < ActiveInteraction::Base
  include ObjectOrId

  object_or_id :user

  def execute
    inputs
  end
end

Example.run!(user: User.new(123))
# => {:user_id=>123, :user=>#<struct User id=123>}
Example.run!(user_id: 123)
# => {:user_id=>123, :user=>#<struct User id=123>}
Example.run!()
# ActiveInteraction::InvalidInteractionError: must give one of user_id or user
Example.run!(user: User.new(123), user_id: 456)
# ActiveInteraction::InvalidInteractionError: can't give both user_id and user
```

**Comment by voondo** [comment]
Maybe a more generic approach could be the ability to manage alternative types for a given input.

e.g:

```
class User < ActiveRecord::Base
end

class Example < ActiveInteraction::Base
  object :user, 
    class: 'User',
    alternatives: ->{
      integer value: ->(input){ User.find(input) }
      object class: GlobalID, value: (input){ input.find }
      string value: ->(input){ GlobalID::Locator.locate input }
    }
end

Example.run!(user: User.find(123))
Example.run!(user: 123)
Example.run!(user: GlobalID.new('gid://app/User/123'))
Example.run!(user: 'gid://app/User/123')
```

What do you think about that ?

**Comment by tfausak** [comment]
Sounds complicated. It's a [union type](https://en.wikipedia.org/wiki/Union_type) with different code running for each type in the union.

**Comment by voondo** [comment]
It's not really an union type because each alternative would produce a value of the main type (no need to keep the original alternative input).
In my previous example, `user` would always be an `User` when called in execute (neither a string, an int or a GlobalID).

Moreover, it would allow code changes without breaking the existing API.
Example:
```ruby
class Example <  ActiveInteraction::Base
  string :name
  def execute
    name.kind_of? String # true
  end
end
# could become
class Example <  ActiveInteraction::Base
  hash :name,
    alternatives: ->{
      string value: ->(input){ { I18n.default_locale => input } }
    } do
      string :en
      string :fr
      string :de
    end
  def execute
    name.kind_of? Hash # true
  end
end
# now we have a I18n name and we didn't break the previous API
# no need to update the X other classes using 'Example'
```

Anyway I often encounter this need and, for me, your solution sounds complicated too :wink: :stuck_out_tongue:

**Comment by tfausak** [comment]
Yeah, the output type is only one type, but the input type is a union. 

I think it's too complex to add in. For example, any of the input types might overlap with another, so we'd have to match them in order. That would require making an interaction's filters ordered, which is not fun. 

I vote to close and tag "wontfix". What do you think, @AaronLasseigne?

**Comment by AaronLasseigne** [comment]
I've certainly run into this situation before. My solution was to wrap the component that accepts a model with one that accepts an ID. This was either done with [`compose`](https://github.com/orgsync/active_interaction/#composition) or by wrapping it in a PORO. The advantage I found was that at times the wrapper might need additional logic. For example, I ran background jobs where it was possible to run a duplicate job. The jobs were idempotent so it wasn't harmful to the data but the extra processing was killing me. Instead, I had the wrapper check the queue and only add the job if a duplicate didn't already exist.

I share @tfausak concern about complicating the interface. We provide mechanisms to introspect the model for building things like APIs. I'm not sure what we'd do in a case like that. As he brought up, certain solutions would force us to consider order. It's certainly convenient. I'm concerned it's a convenience that might not be worth the complication.
